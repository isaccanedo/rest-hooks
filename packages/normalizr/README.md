# normalizr
[![CircleCI](https://circleci.com/gh/coinbase/rest-hooks/tree/master.svg?style=shield)](https://circleci.com/gh/coinbase/rest-hooks)
[![Coverage Status](https://img.shields.io/codecov/c/gh/coinbase/rest-hooks/master.svg?style=flat-square)](https://app.codecov.io/gh/coinbase/rest-hooks?branch=master)
[![npm version](https://img.shields.io/npm/v/@rest-hooks/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/@rest-hooks/normalizr) [![npm downloads](https://img.shields.io/npm/dm/@rest-hooks/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/@rest-hooks/normalizr)

## Install

Install from the NPM repository using yarn or npm:

```shell
yarn add @rest-hooks/normalizr
```

```shell
npm install @rest-hooks/normalizr
```

## Motivation

Many APIs, public or not, return JSON data that has deeply nested objects. Using data in this kind of structure is often very difficult for JavaScript applications, especially those using [Flux](http://facebook.github.io/flux/) or [Redux](http://redux.js.org/).

## Solution

Normalizr is a small, but powerful utility for taking JSON with a schema definition and returning nested entities with their IDs, gathered in dictionaries.

## Documentation

* [Introduction](./docs/README.md)
  * [Build Files](./docs/README.md#build-files)
* [Quick Start](https://resthooks.io/docs/api/schema)
* [API](https://resthooks.io/docs/api/Entity)
  * [normalize](./docs/api.md#normalizedata-schema)
  * [denormalize](./docs/api.md#denormalizeinput-schema-entities)
  * [schema](https://resthooks.io/docs/api/schema)

## Examples

* [Normalizing GitHub Issues](/examples/normalizr-github)
* [Relational Data](/examples/normalizr-relationships)
* [Interactive Redux](/examples/normalizr-redux)

## Quick Start

Consider a typical blog post. The API response for a single post might look something like this:

```json
{
  "id": "123",
  "author": {
    "id": "1",
    "name": "Paul"
  },
  "title": "My awesome blog post",
  "comments": [
    {
      "id": "324",
      "commenter": {
        "id": "2",
        "name": "Nicole"
      }
    }
  ]
}
```

We have two nested entity types within our `article`: `users` and `comments`. Using various `schema`, we can normalize all three entity types down:

```js
import { normalize, schema, Entity } from '@rest-hooks/normalizr';

// Define a users schema
class User extends Entity {
  pk() { return this.id }
}

// Define your comments schema
class Comment extends Entity {
  pk() { return this.id }

  static schema = {
    commenter: User,
  }
}

// Define your article
class Article extends Entity {
  pk() { return this.id }

  static schema = {
  author: User,
  comments: [Comment]
  }
}

const normalizedData = normalize(originalData, Article);
```

Now, `normalizedData` will be:

```js
{
  result: "123",
  entities: {
    "Article": {
      "123": {
        id: "123",
        author: "1",
        title: "My awesome blog post",
        comments: [ "324" ]
      }
    },
    "User": {
      "1": { "id": "1", "name": "Paul" },
      "2": { "id": "2", "name": "Nicole" }
    },
    "Comment": {
      "324": { id: "324", "commenter": "2" }
    }
  }
}
```

## Dependencies

None.

## Credits

Normalizr was originally created by [Dan Abramov](http://github.com/gaearon) and inspired by a conversation with [Jing Chen](https://twitter.com/jingc). Since v3, it was completely rewritten and maintained by [Paul Armstrong](https://twitter.com/paularmstrong).
Since v4, it was largely rewritten and maintained by [Nathaniel Tucker](https://twitter.com/npinp).
It has also received much help, enthusiasm, and contributions from [community members](https://github.com/ntucker/normalizr/graphs/contributors).
