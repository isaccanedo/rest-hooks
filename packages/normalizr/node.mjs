export * from './dist/normalizr.js';
