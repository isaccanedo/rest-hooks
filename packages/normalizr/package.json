{
  "name": "@rest-hooks/normalizr",
  "version": "9.2.0",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://github.com/coinbase/rest-hooks/tree/master/packages/normalizr#readme",
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/normalizr"
  },
  "keywords": [
    "flux",
    "redux",
    "normalize",
    "denormalize",
    "data",
    "cache",
    "api",
    "react-native",
    "ios",
    "android",
    "web",
    "json"
  ],
  "files": [
    "dist/",
    "lib/",
    "node.mjs",
    "legacy",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "sideEffects": false,
  "scripts": {
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development rollup -c",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production rollup -c",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development rollup -c",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:bundle": "run-s build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib ts3.4 legacy dist *.tsbuildinfo",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4",
    "lint": "yarn lint:cmd --fix",
    "lint:ci": "yarn lint:cmd",
    "lint:cmd": "eslint . --ext '.js,.json,.snap' --cache",
    "precommit": "flow check && lint-staged",
    "prepublishOnly": "npm run build"
  },
  "author": "Nathaniel Tucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "devDependencies": {
    "@babel/cli": "7.19.3",
    "@babel/core": "7.20.2",
    "@types/babel__core": "^7",
    "downlevel-dts": "^0.10.0",
    "immutable": "4.1.0",
    "npm-run-all": "^4.1.5",
    "rollup": "2.79.1",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "npm-run-all": "^4.1.5"
  }
}
