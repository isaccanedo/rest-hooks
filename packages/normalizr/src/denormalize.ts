import type { Schema, EntityInterface, UnvisitFunction } from './interface.js';
import type {
  Denormalize,
  DenormalizeNullable,
  DenormalizeCache,
} from './types.js';
import { isEntity } from './isEntity.js';
import WeakListMap from './WeakListMap.js';
import { denormalize as arrayDenormalize } from './schemas/Array.js';
import { denormalize as objectDenormalize } from './schemas/Object.js';
import { isImmutable } from './schemas/ImmutableUtils.js';

const DRAFT = Symbol('draft');

const unvisitEntity = (
  entityOrId: Record<string, any> | string,
  schema: EntityInterface,
  unvisit: UnvisitFunction,
  getEntity: (
    entityOrId: Record<string, any> | string,
    schema: EntityInterface,
  ) => object | symbol,
  localCache: Record<string, Record<string, any>>,
  entityCache: DenormalizeCache['entities'],
  dependencies: object[],
  cycleIndex: { i: number },
): [denormalized: object | undefined, found: boolean, deleted: boolean] => {
  const entity = getEntity(entityOrId, schema);
  if (
    typeof entity === 'symbol' &&
    (entity as symbol).toString().includes('DELETED')
  ) {
    return [undefined, true, true];
    // TODO: Change to this as breaking change once we only support newer entities
    // also remove `(entity as symbol).toString().includes('DELETED')` and do this for all symbols
    // return schema.denormalize(entity, unvisit);
  }

  if (typeof entity !== 'object' || entity === null) {
    return [entity as any, false, false];
  }

  const pk =
    typeof entityOrId === 'string'
      ? entityOrId
      : schema.pk(isImmutable(entity) ? (entity as any).toJS() : entity);
  if (pk === undefined || pk === '') {
    return [entity, false, false];
  }

  if (localCache[schema.key] === undefined) {
    localCache[schema.key] = Object.create(null);
  }

  let found = true;
  let deleted = false;

  if (!localCache[schema.key][pk]) {
    const trackingIndex = dependencies.length;
    dependencies.push(entity);

    const wrappedUnvisit = withTrackedEntities(unvisit);
    // { [DRAFT] } means we are still processing - which if found indicates a cycle
    wrappedUnvisit.setLocal = entityCopy =>
      (localCache[schema.key][pk] = { [DRAFT]: entityCopy, i: trackingIndex });

    const globalCacheEntry = getGlobalCacheEntry(entityCache, schema, pk);

    [localCache[schema.key][pk], found, deleted] = schema.denormalize(
      entity,
      wrappedUnvisit,
    );

    // if in cycle, use the start of the cycle to track all deps
    // otherwise, we use our own trackingIndex
    const localKey = dependencies.slice(
      cycleIndex.i === -1 ? trackingIndex : cycleIndex.i,
    );

    if (!globalCacheEntry.has(localKey)) {
      globalCacheEntry.set(localKey, localCache[schema.key][pk]);
    } else {
      localCache[schema.key][pk] = globalCacheEntry.get(localKey);
    }

    // start of cycle - reset cycle detection
    if (cycleIndex.i === trackingIndex) {
      cycleIndex.i = -1;
    }
  } else {
    // cycle detected
    if (Object.hasOwn(localCache[schema.key][pk], DRAFT)) {
      cycleIndex.i = localCache[schema.key][pk].i;
      return [localCache[schema.key][pk][DRAFT], found, deleted];
    } else {
      // with no cycle, globalCacheEntry will have already been set
      dependencies.push(entity);
    }
  }

  return [localCache[schema.key][pk], found, deleted];
};

const getUnvisit = (
  entities: Record<string, Record<string, any>>,
  entityCache: DenormalizeCache['entities'],
  resultCache: WeakListMap<object, any>,
  localCache: Record<string, Record<string, any>>,
) => {
  const getEntity = getEntities(entities);
  const dependencies: object[] = [];
  const cycleIndex = { i: -1 };

  function unvisit(
    input: any,
    schema: any,
  ): [denormalized: any, found: boolean, deleted: boolean] {
    if (!schema) return [input, true, false];

    // null is considered intentional, thus always 'found' as true
    if (input === null) {
      return [input, true, false];
    }

    const hasDenormalize = typeof schema.denormalize === 'function';

    if (!hasDenormalize && typeof schema === 'function') {
      if (input instanceof schema) return [input, true, false];
      // field deserialization should never count against 'found' (whether to used inferred results)
      if (input === undefined) return [input, true, false];
      return [new schema(input), true, false];
    }

    if (input === undefined) {
      return [input, false, false];
    }

    if (!hasDenormalize && typeof schema === 'object') {
      const method = Array.isArray(schema)
        ? arrayDenormalize
        : objectDenormalize;
      return method(schema, input, unvisit);
    }

    if (isEntity(schema)) {
      return unvisitEntity(
        input,
        schema,
        unvisit,
        getEntity,
        localCache,
        entityCache,
        dependencies,
        cycleIndex,
      );
    }

    if (hasDenormalize) {
      return schema.denormalize(input, unvisit);
    }

    return [input, true, false];
  }

  //const wrappedUnvisit = withTrackedEntities(unvisit, globalKey);

  return (
    input: any,
    schema: any,
  ): [denormalized: any, found: boolean, deleted: boolean] => {
    const ret = unvisit(input, schema);
    // in the case where WeakMap cannot be used
    // this test ensures null is properly excluded from WeakMap
    if (Object(input) !== input) return ret;

    dependencies.push(input);
    if (!resultCache.has(dependencies)) {
      resultCache.set(dependencies, ret[0]);
      return ret;
    } else {
      return [resultCache.get(dependencies), ret[1], ret[2]];
    }
  };
};

const getEntities = (entities: Record<string, any>) => {
  const entityIsImmutable = isImmutable(entities);

  return (
    entityOrId: Record<string, any> | string,
    schema: EntityInterface,
  ) => {
    const schemaKey = schema.key;

    if (typeof entityOrId === 'object') {
      return entityOrId;
    }

    if (entityIsImmutable) {
      return entities.getIn([schemaKey, entityOrId]);
    }

    return entities[schemaKey]?.[entityOrId];
  };
};

type DenormalizeReturn<S extends Schema> =
  | [
      denormalized: Denormalize<S>,
      found: true,
      deleted: false,
      resolvedEntities: Record<string, Record<string, any>>,
    ]
  | [
      denormalized: DenormalizeNullable<S>,
      found: boolean,
      deleted: true,
      resolvedEntities: Record<string, Record<string, any>>,
    ]
  | [
      denormalized: DenormalizeNullable<S>,
      found: false,
      deleted: boolean,
      resolvedEntities: Record<string, Record<string, any>>,
    ];

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const denormalize = <S extends Schema>(
  input: unknown,
  schema: S | undefined,
  entities: any,
  entityCache: DenormalizeCache['entities'] = {},
  resultCache: WeakListMap<object, any> = new WeakListMap(),
): DenormalizeReturn<S> => {
  // undefined mean don't do anything
  if (schema === undefined) {
    return [input, true, false, {}] as [any, boolean, boolean, any];
  }
  if (input === undefined) {
    return [undefined, false, false, {}] as [any, boolean, boolean, any];
  }
  const resolvedEntities: Record<string, Record<string, any>> = {};
  const unvisit = getUnvisit(
    entities,
    entityCache,
    resultCache,
    resolvedEntities,
  );
  return [...unvisit(input, schema), resolvedEntities] as [
    any,
    boolean,
    boolean,
    any,
  ];
};

export const denormalizeSimple = <S extends Schema>(
  input: any,
  schema: S | undefined,
  entities: any,
  entityCache: DenormalizeCache['entities'] = {},
  resultCache: WeakListMap<object, any> = new WeakListMap(),
):
  | [denormalized: Denormalize<S>, found: true, deleted: false]
  | [denormalized: DenormalizeNullable<S>, found: boolean, deleted: true]
  | [denormalized: DenormalizeNullable<S>, found: false, deleted: boolean] =>
  denormalize(input, schema, entities, entityCache, resultCache).slice(
    0,
    3,
  ) as any;

function getGlobalCacheEntry(
  entityCache: {
    [key: string]: { [pk: string]: WeakListMap<object, EntityInterface<any>> };
  },
  schema: any,
  id: any,
) {
  if (!entityCache[schema.key]) entityCache[schema.key] = {};
  if (!entityCache[schema.key][id])
    entityCache[schema.key][id] = new WeakListMap();
  return entityCache[schema.key][id];
}

function withTrackedEntities(unvisit: UnvisitFunction): UnvisitFunction {
  // every time we nest, we want to unwrap back to the top.
  // this is due to only needed the next level of nested entities for lookup
  const originalUnvisit = unvisit.og || unvisit;
  const wrappedUnvisit = (input: any, schema: any) =>
    originalUnvisit(input, schema);
  wrappedUnvisit.og = unvisit;
  return wrappedUnvisit;
}
