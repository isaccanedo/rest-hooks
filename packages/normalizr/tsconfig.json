{
  "extends": "../../tsconfig-base",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src",
    "noImplicitAny": false,
    "skipLibCheck": false
  },
  "include": ["src"]
}
