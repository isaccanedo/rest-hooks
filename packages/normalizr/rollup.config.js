import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import filesize from 'rollup-plugin-filesize';
import resolve from 'rollup-plugin-node-resolve';
import { terser } from 'rollup-plugin-terser';
import dts from 'rollup-plugin-dts';

import { name } from './package.json';

const isProduction = process.env.NODE_ENV === 'production';

const destBase = 'dist/normalizr';
const destExtension = `${isProduction ? '.min' : ''}.js`;
const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node'];

function snakeCase(id) {
  return id.replace(/(-|\/)/g, '_').replace(/@/g, '');
}

const configs = [];

if (process.env.BROWSERSLIST_ENV !== 'node12') {
  if (process.env.NODE_ENV === 'production') {
    configs.push({
      input: 'lib/index.js',
      output: [{ file: `${destBase}.es${destExtension}`, format: 'es' }],
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          runtimeHelpers: true,
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
  } else {
    configs.push({
      input: 'lib/index.js',
      output: [
        {
          file: `${destBase}.umd${destExtension}`,
          format: 'umd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.amd${destExtension}`,
          format: 'amd',
          name: snakeCase(name),
        },
        {
          file: `${destBase}.browser${destExtension}`,
          format: 'iife',
          name: snakeCase(name),
        },
      ],
      plugins: [
        babel({
          exclude: ['node_modules/**', '/**__tests__/**'],
          extensions,
          runtimeHelpers: true,
          rootMode: 'upward',
        }),
        resolve({ extensions }),
        commonjs({ extensions }),
        isProduction && terser(),
        filesize(),
      ].filter(Boolean),
    });
    configs.push({
      input: './lib/index.d.ts',
      output: [{ file: 'index.d.ts', format: 'es' }],
      plugins: [dts()],
    });
  }
} else {
  // node-friendly commonjs build
  configs.push({
    input: 'lib/index.js',
    output: [{ file: `${destBase}${destExtension}`, format: 'cjs' }],
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        runtimeHelpers: true,
        rootMode: 'upward',
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      isProduction && terser(),
      filesize(),
    ].filter(Boolean),
  });
}
export default configs;
