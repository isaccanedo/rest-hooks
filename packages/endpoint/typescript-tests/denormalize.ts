import { normalize, denormalize } from '@rest-hooks/normalizr';
import { IDEntity } from '__tests__/new';

import { schema, DenormalizeNullable, Normalize, Denormalize } from '../src';

class Magic extends IDEntity {
  readonly a = 'first' as const;
  private b = false;
  c = 'arg' as const;
}
class Magic2 extends IDEntity {
  readonly a = 'second' as const;
  private b3 = false;
}
const magicSchema = Magic;
const magic2Schema = Magic2;

const unionSchema = new schema.Union(
  {
    user: magicSchema,
    group: magic2Schema,
  },
  'a',
);
const errorUnionSchema = new schema.Union(
  {
    user: magicSchema,
    group: magic2Schema,
  },
  // @ts-expect-error
  'notexistant',
);
const scheme = {
  thing: { data: unionSchema, members: [Magic] },
  first: '',
  second: '',
};
const schemeEntity = Magic;

const de = denormalize({}, scheme, {});
const r = normalize({}, scheme);

type A = DenormalizeNullable<typeof scheme>;
type B = A['thing']['members'];
type C = DenormalizeNullable<typeof schemeEntity>;
type D = ReturnType<typeof unionSchema['_denormalizeNullable']>;
type F = Denormalize<typeof unionSchema>;
type E = Normalize<typeof scheme>['thing']['data'];

if (de[1]) {
  const value = de[0];
  const piece = value.thing.data?.a;
  const first: string = value.first;
  const members = value.thing.members;
} else {
  const value = de[0];
  const data = value.thing.data;
  const members = value.thing.members;
}
const members2 = de[0].thing.members;

const schemeValues = new schema.Values({ btc: Magic, eth: Magic2 });
const schemeValuesSimple = new schema.Values(Magic);
const [valueValues, foundValues] = denormalize({}, schemeValues, {});
Object.keys(schemeValues).forEach(k => {
  const v = valueValues[k];
  if (v?.a === 'second') {
    const b: Magic2 = v;
  }
});

const [valueValuesSimple, foundValuesSimple] = denormalize(
  {},
  schemeValuesSimple,
  {},
);
