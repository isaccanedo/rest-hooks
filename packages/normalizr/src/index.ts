import { denormalize } from './denormalize.js';
import { normalize } from './normalize.js';
import WeakListMap from './WeakListMap.js';
import { isEntity } from './isEntity.js';
export { default as inferResults } from './inferResults.js';
export { DELETED } from './special.js';

export type {
  AbstractInstanceType,
  NormalizeReturnType,
  NormalizedSchema,
  DenormalizeReturnType,
  DenormalizeCache,
} from './types.js';
export * from './endpoint/types.js';
export * from './interface.js';
export * from './Expiry.js';
export * from './normal.js';

export { denormalize, normalize, isEntity, WeakListMap };
