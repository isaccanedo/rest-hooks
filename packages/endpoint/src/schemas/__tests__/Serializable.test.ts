// eslint-env jest
import { normalize } from '@rest-hooks/normalizr';
import { IDEntity } from '__tests__/new';

import denormalize from './denormalize';
import { schema } from '../../';

let dateSpy;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

class User extends IDEntity {
  createdAt = new Date(0);
  name = '';
  static schema = {
    createdAt: Date,
  };
}
class Other {
  thing = 0;
  constructor(props) {
    this.thing = props.thing;
  }

  toJSON() {
    return { thing: this.thing };
  }
}
const objectSchema = {
  user: User,
  anotherItem: Other,
  time: Date,
};

describe(`Serializable normalization`, () => {
  test('normalizes date and custom', () => {
    const norm = normalize(
      {
        user: {
          id: '1',
          name: 'Nacho',
          createdAt: '2020-06-07T02:00:15+0000',
        },
        anotherItem: { thing: 500 },
        time: '2020-06-07T02:00:15+0000',
      },
      objectSchema,
    );
    expect(norm.result.time.getTime()).toBe(norm.result.time.getTime());
    expect(norm.result.anotherItem).toBeInstanceOf(Other);
    expect(norm.entities[User.key]['1'].createdAt).toBe(
      norm.entities[User.key]['1'].createdAt,
    );
    expect(norm).toMatchSnapshot();
    expect(JSON.stringify(norm)).toMatchSnapshot();
  });
});

describe(`Serializable denormalization`, () => {
  test('denormalizes date and custom', () => {
    const entities = {
      User: {
        '1': {
          id: '1',
          name: 'Nacho',
          createdAt: new Date('2020-06-07T02:00:15+0000'),
        },
      },
    };
    const [response, found] = denormalize(
      {
        user: '1',
        anotherItem: new Other({ thing: 500 }),
        time: new Date('2020-06-07T02:00:15+0000'),
      },
      objectSchema,
      entities,
    );
    expect(response.anotherItem).toBeInstanceOf(Other);
    expect(response.time.getTime()).toBe(response.time.getTime());
    expect(response.user?.createdAt.getTime()).toBe(
      response.user?.createdAt.getTime(),
    );
    expect(found).toBe(true);
    expect(response).toMatchSnapshot();
  });

  test('denormalizes as plain', () => {
    const entities = {
      User: {
        '1': {
          id: '1',
          name: 'Nacho',
          createdAt: '2020-06-07T02:00:15+0000',
        },
      },
    };
    const [response, found] = denormalize(
      {
        user: '1',
        anotherItem: { thing: 500 },
        time: '2020-06-07T02:00:15+0000',
      },
      objectSchema,
      entities,
    );
    expect(response.anotherItem).toBeInstanceOf(Other);
    expect(response.time.getTime()).toBe(response.time.getTime());
    expect(response.user?.createdAt.getTime()).toBe(
      response.user?.createdAt.getTime(),
    );
    expect(found).toBe(true);
    expect(response).toMatchSnapshot();
  });
});
