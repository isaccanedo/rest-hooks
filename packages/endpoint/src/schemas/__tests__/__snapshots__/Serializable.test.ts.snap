// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serializable denormalization denormalizes as plain 1`] = `
{
  "anotherItem": {
    "thing": 500,
  },
  "time": 2020-06-07T02:00:15.000Z,
  "user": User {
    "createdAt": 2020-06-07T02:00:15.000Z,
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`Serializable denormalization denormalizes date and custom 1`] = `
{
  "anotherItem": {
    "thing": 500,
  },
  "time": 2020-06-07T02:00:15.000Z,
  "user": User {
    "createdAt": 2020-06-07T02:00:15.000Z,
    "id": "1",
    "name": "Nacho",
  },
}
`;

exports[`Serializable normalization normalizes date and custom 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "createdAt": 2020-06-07T02:00:15.000Z,
        "id": "1",
        "name": "Nacho",
      },
    },
  },
  "entityMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "anotherItem": {
      "thing": 500,
    },
    "time": 2020-06-07T02:00:15.000Z,
    "user": "1",
  },
}
`;

exports[`Serializable normalization normalizes date and custom 2`] = `"{"entities":{"User":{"1":{"id":"1","name":"Nacho","createdAt":"2020-06-07T02:00:15.000Z"}}},"indexes":{},"result":{"user":"1","anotherItem":{"thing":500},"time":"2020-06-07T02:00:15.000Z"},"entityMeta":{"User":{"1":{"expiresAt":null,"date":1557831718135,"fetchedAt":0}}}}"`;
